### PR TITLE
payment link end point

### PR DESCRIPTION
**Pull request by neera11**

### CODE REVIEW

**Comment by isaumya** [comment]
Hey @neera11,
so if I am using this Razorpay PHP client to greater standard Payment Links, so, do I have to change the structure of the data that is passed? For example, current as per the readme example, it is stayed that I have to call the client the following way to create a payment link:

```php
$link  = $api->invoice->create(
	arary(
		'type' => 'link', 
		'amount' => 500, 
		'description' => 'For XYZ purpose', 
		'customer' => array(
			'email' => 'test@test.test'
		)
	)
);
```

So, do I have to make any changes to this structure if I am only generating standard payment links?

**Comment by neera11** [comment]
Hi @isaumya 
No need to pass type in Payment link call
with invoice we were passing type as "link"

For payment link data will be 

$orderData = json_encode(
    [
    'amount' => 98765, 
    'description' => 'For XYZ purpose', 
    'customer' => array('email' => 'test@test.test')
    ]);

$link = $api->payment_link->create($orderData, 'application/json' );

**Comment by isaumya** [comment]
Hi @neera11,
Thanks for your reply. So, no we need to pass `application/json` as a second parameter to the `create()` function? As this was not there before.

I will highly recommend you to update the `readme.md` file with example of the new way of calling the API.

So, just to be clear, currently, I am doing the following for generating payment link:

```php
// Include the composer autoload file
require_once( __DIR__ . '/../vendor/autoload.php' );

// Use Razorpay API
use Razorpay\Api\Api;

/**
 * Function to generate Razorpay Payment Link
 * @var $payment_dtls Array
 * @return $link Object
 */
function generate_payment_link( $payment_dtls ) {
  // Initiate New API
  $api = new Api( 'rzp_live_123456', 'abc123' );

  // Create Payment Link
  $link = $api->invoice->create(
    [
      'type'            =>  'link',
      'amount'          =>  ( $payment_dtls['amount'] * 100 ),
      'currency'        =>  $payment_dtls['currency'],
      'customer_id'     =>  $payment_dtls['cust_id'], // Customer ID
      'description'     =>  $payment_dtls['description'],
      'reminder_enable' =>  true,
      'expire_by'       =>  strtotime( $payment_dtls['expiry'] ) // Payment link will expire in X time from generation 
    ]
  );

  // Return the payment link details send by Razorpay after payment link creation
  return $link;
}
```

and once this new version of the package is released, it would done as following:

```php
// Include the composer autoload file
require_once( __DIR__ . '/../vendor/autoload.php' );

// Use Razorpay API
use Razorpay\Api\Api;

/**
 * Function to generate Razorpay Payment Link
 * @var $payment_dtls Array
 * @return $link Object
 */
function generate_payment_link( $payment_dtls ) {
  // Initiate New API
  $api = new Api( 'rzp_live_123456', 'abc123' );

  // Create Payment Link
  $link = $api->payment_link->create(
    json_encode(
      [
        'amount'          =>  ( $payment_dtls['amount'] * 100 ),
        'currency'        =>  $payment_dtls['currency'],
        'customer_id'     =>  $payment_dtls['cust_id'], // Customer ID
        'description'     =>  $payment_dtls['description'],
        'reminder_enable' =>  true,
        'expire_by'       =>  strtotime( $payment_dtls['expiry'] ) // Payment link will expire in X time from generation 
      ]
    ),
    'application/json'
  );

  // Return the payment link details send by Razorpay after payment link creation
  return $link;
}
```

Right? Or I am missing/misunderstanding something here?

**Comment by neera11** [comment]
@isaumya 
You are using correctly
Is it woking fine for you
or you getting any error?

**Comment by isaumya** [comment]
Hi @neera11,
I haven't tested the above code, I just modified the code to show you. I am not sure if I can test it until you guys release the update and I update the package via composer. I've running v2.7.0

**Comment by isaumya** [comment]
Hey @neera11,
You still haven't updated the readme file which still suggests users to create the payment link in the old ways.

**Comment by neera11** [comment]
Hi @isaumya 
sorry for the delay
I have updated readme file. Please check

**Comment by isaumya** [comment]
Thanks, @neera11. Yes, I have checked. Now I am waiting for this PR to be accepted and pushing a new release with this code so that I can update the dependency and use the new methods.

**Comment by isaumya** [comment]
Hey @neera11,
When is this getting released in a new version of the PHP client? As we have to update to the new endpoint by Aug 15, right? So, not many days left.

File: src/Entity.php [hunk 1]
```diff
@@ -7,12 +7,18 @@
...
-
-    protected function create($attributes = null)
+ /**
+     * Creates a payment link 
+     *
+     * @param array $attributes 
+     * @param string $contentType
+     * @return Payment_link
+     */
+    protected function create($attributes = null, $contentType= 'application/x-www-form-urlencoded')
```
[the remark this answers is not in the document]
Comment (abhayp) [reply src/Entity.php:17]: I dont think clients should pass content type separately. Idea of sdk is using the functions without knowing the internals. it should be just paymentlink->create(attributes) imo. content type if required can be passed internally to the generic function

File: src/Request.php [hunk 2]
```diff
@@ -48,13 +48,11 @@ public function request($method, $url, $data = array())
...
+            'timeout' => 60
         );
-
-        $headers = $this->getRequestHeaders();
-
-        $response = Requests::request($url, $headers, $data, $method, $options);
-
+        
+        $headers = $this->getRequestHeaders($contentType);
+        $response = Requests::request($url, $headers, $data, $method, $options);
```
[the remark this answers is not in the document]
Comment (abhayp) [reply src/Request.php:56]: yes

File: src/Request.php [hunk 3]
```diff
@@ -153,12 +151,17 @@ protected function throwServerError($body, $httpStatusCode)
             $httpStatusCode);
     }
 
-    protected function getRequestHeaders()
+    protected function getRequestHeaders($contentType = 'application/x-www-form-urlencoded')
```
[the remark this answers is not in the document]
Comment (abhayp) [reply src/Request.php:155]: What was the default content type before? if we start passing default content type as above for all apis, will there be any issues for other functionalities?

File: src/Entity.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhayp) [reply src/Entity.php:17]: @chirag-patel-rp

[the remark this answers is not in the document]
Comment (chirag-patel-rp) [reply src/Entity.php:17]: @neera11 yes we should not explicitly pass content type. The idea of SDK is to abstract out common things and only allow users to make changes in the input values, also check other APIs implemented, I don’t see anywhere we passed explicitly content type="application/json"

File: src/Request.php [hunk 4]
```diff
@@ -153,12 +151,17 @@ protected function throwServerError($body, $httpStatusCode)
...
     {
+        
+            $ContentTypeHeader = array('Content-Type' => $contentType );
+            self::$headers = array_merge(self::$headers, $ContentTypeHeader);
```
[the remark this answers is not in the document]
Comment (chirag-patel-rp) [reply src/Request.php:160]: @abhayp @sofiya02 can we ask each developer to use the same code editor and have some linters in place, to avoid such issues in the code review

**Comment by isaumya** [comment]
@abhayp, @chirag-patel-rp please hurry up on pushing this release to the SDK as my account has been upgraded to using the new API and all of my automation system which is using the SDK might fail as I am not using the new API for payment link generation.

**Comment by chirag-patel-rp** [comment]
@isaumya contenty type relate comment is still not resolved
@neera11 any update on that?

**Comment by neera11** [comment]
@chirag-patel-rp @isaumya  I have added content type related issue

File: src/Request.php [hunk 5]
```diff
@@ -153,12 +152,18 @@ protected function throwServerError($body, $httpStatusCode)
             $httpStatusCode);
     }
 
-    protected function getRequestHeaders()
+    protected function getRequestHeaders($additionHeader)
     {
+        if(isset($additionHeader)){
```
Comment (abhayp) [inline src/Request.php:157]: dont check isset. it will always return true since the variable is set. check if(empty(additionHeader) === false please

File: src/Request.php [hunk 6]
```diff
@@ -153,12 +152,18 @@ protected function throwServerError($body, $httpStatusCode)
             $httpStatusCode);
     }
 
-    protected function getRequestHeaders()
+    protected function getRequestHeaders($additionHeader)
```
Comment (abhayp) [inline src/Request.php:155]: add param type. this should always be an array. empty array for non pl

File: src/Entity.php [hunk 7]
```diff
@@ -7,12 +7,19 @@
...
+     *
+     * 
+     */
     protected function create($attributes = null)
     {
         $entityUrl = $this->getEntityUrl();
-
-        return $this->request('POST', $entityUrl, $attributes);
+        
+        return $this->request('POST', $entityUrl, $attributes, $this->additionHeader);
```
Comment (abhayp) [inline src/Entity.php:22]: additional header is passed here. cannot see the function param in request function

File: src/Entity.php [hunk 8]
```diff
@@ -83,10 +90,9 @@ protected function request($method, $relativeUrl, $data = null)
     {
         $request = new Request();
 
-        $response = $request->request($method, $relativeUrl, $data);
+        $response = $request->request($method, $relativeUrl, $data, $this->additionHeader);
```
Comment (abhayp) [inline src/Entity.php:93]: why pass the additional header from create function if u are referencing this here

File: src/Request.php [hunk 9]
```diff
@@ -37,7 +37,7 @@ class Request
      * @return array Response data in array format. Not meant
      * to be used directly
      */
-    public function request($method, $url, $data = array())
+    public function request($method, $url, $data = array(),$additionHeader)
```
Comment (abhayp) [inline src/Request.php:40]: initialize as empty array please ..just like data

**Review by abhayp — APPROVED** [review]
lgtm

File: README.md [hunk 10]
```diff
@@ -74,10 +74,20 @@ $transfer  = $api->transfer->fetch($transferId)->edit($options); // Edit a trans
...
 
 // Payment Links
-$links = $api->invoice->all();
-$link  = $api->invoice->fetch('inv_00000000000001');
-$link  = $api->invoice->create(arary('type' => 'link', 'amount' => 500, 'description' => 'For XYZ purpose', 'customer' => array('email' => 'test@test.test')));
+$links = $api->payment_link->all(); // fetch all payment links
+
+$link  = $api->payment_link->fetch('plink_GiwM9xbIZqbkJp'); // fetch payment link with id
+
+$data = json_encode(
```
Comment (ramth05) [inline README.md:81]: @neera11, why we need to pass data as JSON for this entity only?

File: src/Payment_link.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Razorpay\Api;
+
+use Requests;
+
+
+class Payment_link extends Entity
```
Comment (ramth05) [inline src/Payment_link.php:8]: @neera11 , why it's `Payment_link` not `PaymentLink`

File: src/Request.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (ramth05) [reply src/Request.php]: @abhayp, All of other RZP API is accepting POST request with `Content-Type: application/x-www-form-urlencoded` then why payment Links not ?

[the remark this answers is not in the document]
Comment (abhayp) [reply src/Request.php]: all out apis are json rest apis. we shouldnt ideally support  application/x-www-form-urlencoded for json request. but since legacy code base is like that, we cannot remove it. for new ones we should enforce application/json

File: README.md (hunk 10 above)
Comment (abhayp) [reply README.md:81]: yep..should be array. @neera11

File: src/Request.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (chirag-patel-rp) [reply src/Request.php]: we can release the new SDK version with `application/json` @ramth05 do you see any challenges or drawbacks?
Anyway we want to move to `application/json`

[the remark this answers is not in the document]
Comment (ramth05) [reply src/Request.php]: @chirag-patel-rp , don't see any issue as all our API supporting `application/json` and no changes required by enduser . The only thing is that all the API end points we should test it before release new version. and if there is any corner case it should be handled.

**Comment by neera11** [comment]
Hi @isaumya 

We have updated the paymentlink.
Now you don't need to pass second parameter  as application/json while creating new paymentlink and also no need to pass data in json format.
just use as below
$link = $api->payment_link->create(array('amount' => 98765,'description' => 'For XYZ purpose', 'customer' => array('email' => 'test@test.test'))); // create payment link

Thank you

**Comment by isaumya** [comment]
Hi @neera11,
Thank you so much for letting me know and also for releasing the update.
Just have one small request, in the readme file for each call, please mention the respective API doc link for that call so that users can easily see what are the list of arguments that they can pass to that call. As in your example, you may have used a few parameters while I might use parameters like currency, cusermeID etc. So, if you guys provide a link to the API docs, users can easily see what they can pass to that call.

P.S.; Also in a future release, try testing your codebase in PHP 8, there are a few minor warnings in PHP 8 for the codebase.
